#########################################################################################################    
tabPanel(
  "Manual",
  h1(HTML('<font color="tomato" face="Comic Sans MS"><center><b>Manual</b></center></font>')),
  
  rnaseq2g.write.prompt(c('RNA-seq 2G is a web portal with over 25 statistical methods that perform two-group analysis of differential gene expression. It uses read count data from RNA-seq or similar data matrix as input and generates test statistics in consistent format as output.')),
  
  div(
    style="padding: 0.5cm;", 
    tabsetPanel(
      tabPanel(
        "Introduction",
        HTML('&nbsp;<div style="color:darkblue; padding: 0cm 0.5cm">',
             'Two-group comparison of differential expression (DE) is the most common analysis of transcriptome data. For RNA-seq data,',
             'the comparison is usually performed on a gene-level matrix of read counts,',
             'with the read counts corresponding to the number of sequencing reads mapped to each gene in each RNA-seq sample.',
             '</div>&nbsp;'), 
        div(style="padding: 0.5cm", tabsetPanel(
          tabPanel('DE methods', HTML('&nbsp;'), uiOutput('manual.1.1')),
          tabPanel('Test statistics', HTML('&nbsp;'), uiOutput('manual.1.2')))
        )
      ),
      tabPanel(
        "Prepare inputs",
        HTML('&nbsp;<div style="color:darkblue; padding: 0cm 0.5cm">',
             'For RNA-seq data the raw sequencing reads need to be aligned to the reference genome and transcriptome, using any alignment program.',
             'Next, the aligned reads should be assigned to annotated genes or transcripts to generate a read count matrix.',
             'RNA-seq 2G accepts other types of data, such as those generated by the proteomics and Nanostring technologies,',
             'as long as the raw data was processed similarly to generate a integer matrix.',
             '</div>&nbsp;'),
        div(style="padding: 0.5cm", tabsetPanel(
          tabPanel('Read count matrix', HTML('&nbsp;'), uiOutput('manual.2.1')),
          tabPanel('Sample grouping', HTML('&nbsp;'), uiOutput('manual.2.2')),
          tabPanel('Other parameters', HTML('&nbsp;'), uiOutput('manual.2.3')))
        )
      ),
      tabPanel(
        "Run analysis online",
        HTML('&nbsp;<div style="color:darkblue; padding: 0cm 0.5cm">',
             '<b>RNAseq 2G</b> provides a user-friendly web portal to run a DE analysis using any of the available methods.',
             'Each analysis will be assigned a random ID and its results can be re-visited by specifying the ID.',
             'To perform an analysis, go to <a href="http://rnaseq2g.awsomics.org">rnaseq2g.awsomics.org</a>',
             'and set it up with the following 3 steps.', 
             '</div>&nbsp;'),
        div(style="padding: 0.5cm", tabsetPanel(
          tabPanel('Step 1', HTML('&nbsp;'), uiOutput('manual.3.1')),
          tabPanel('Step 2', HTML('&nbsp;'), uiOutput('manual.3.2')),
          tabPanel('Step 3', HTML('&nbsp;'), uiOutput('manual.3.3')),
          tabPanel('Submit analysis', HTML('&nbsp;'), uiOutput('manual.3.4')))
        )
      ),
      tabPanel(
        "Run analysis offline",
        HTML('&nbsp;<div style="color:darkblue; padding: 0cm 0.5cm">',
             'An alternative to use <b>RNA-seq 2G</b> is to directly call the <b><i>DeRNAseq {DEGandMore}</i></b> function within R.',
             'This option is more suitable for DE analysis runs using any of the slow methods (see <b>Table 1</b>).',
             'When any slow DE methods are selected, the online waiting might be too long and users should run the DE analysis offline,',
             'but can later upload their results to <b>RNAseq 2G</b> for visualization.',
             'Running the offline analysis takes some basic R skills and a few simple steps.',
             '</div>&nbsp;'),
        div(style="padding: 0.5cm", tabsetPanel(
          tabPanel('Install packages', HTML('&nbsp;'), uiOutput('manual.4.1')),
          tabPanel('Run analysis', HTML('&nbsp;'), uiOutput('manual.4.2')),
          tabPanel('List methods', HTML('&nbsp;'), uiOutput('manual.4.3')))
        )
      ),
      tabPanel(
        "Browse results", HTML('&nbsp;'), 
        div(
          style="padding: 0.5cm", 
          tabsetPanel(
            tabPanel('Load results', HTML('&nbsp;'), uiOutput('manual.5.1')),
            tabPanel('View results', HTML('&nbsp;'), uiOutput('manual.5.2'))
          )
        )
      ),
      tabPanel(
        "Compare methods", HTML('&nbsp;'), 
        div(
          style="padding: 0.5cm", 
          tabsetPanel(
            tabPanel('Global pattern', HTML('&nbsp;'), uiOutput('manual.6.1')),
            tabPanel('Compare p values', HTML('&nbsp;'), uiOutput('manual.6.2')),
            tabPanel('Single gene', HTML('&nbsp;'), uiOutput('manual.6.3'))
          )
        )
      ),
      tabPanel(
        "Meta-analysis", HTML('&nbsp;'), 
        div(
          style="padding: 0.5cm", 
          tabsetPanel(
            tabPanel('Set up an analysis', HTML('&nbsp;'), uiOutput('manual.7.1')),
            tabPanel('View results', HTML('&nbsp;'), uiOutput('manual.7.2')),
            tabPanel('Single gene', HTML('&nbsp;'), uiOutput('manual.7.3'))
          )
        )
      )
    )
  ) # END of main manual tabset
)  
